### PR TITLE
Decoder trace generation

### PR DESCRIPTION
**Pull request by bobbinth**
This PR implements decoder trace generation and related changes. This includes:

- [x] Implement trace generation for most columns the decoder component. Trace for 5 more decoder columns is left for another PR. These columns do not affect decoder logic and will be mostly used for degree reductions.
- [x] Add `merge()` and `hash_span_block()` methods to the hasher component. (will need to create an issue for adding tests for these functions).
- [x] Update hasher trace generation so that row address always starts at 1. (will need to create an issue for adding a boundary constraint for this).
- [x] Implement tests for the decoder trace.

### CODE REVIEW

File: core/src/hasher/mod.rs [hunk 1]
```diff
@@ -44,6 +48,9 @@ pub const NUM_ROUNDS: usize = Hasher::NUM_ROUNDS;
 /// Number of selector columns in the trace.
 pub const NUM_SELECTORS: usize = 3;
 
+/// Number of execution trace rows needed to fully describe one permutation of the hash function.
+pub const TRACE_CYCLE_LEN: usize = NUM_ROUNDS + 1;
```
Comment (grjte) [inline core/src/hasher/mod.rs:52]: I find this name a bit confusing. Why not call it HASH_CYCLE_LEN as we do in other places?

**Review by grjte — COMMENTED** [review]
I haven't been able to finish reviewing everything yet, so I'm leaving a partial review for now with a few comments inline. I'll finish the full review tomorrow

File: processor/src/decoder/trace.rs [hunk 2]
```diff
@@ -0,0 +1,374 @@
...
+        // for unfilled rows of hasher state columns, copy over values from the last row for the
+        // first 4 columns, and pad the other 4 columns with ZEROs
+        for (i, mut column) in self.hasher_trace.into_iter().enumerate() {
+            debug_assert_eq!(own_len, column.len());
+            if i < 4 {
+                // TODO: ideally, we should use `expect()` instead of `unwrap_or()` but aux_table
+                // unit tests fail if we do that as they don't expect build the decoder portion of
+                // the trace. in the future we should refactor aux_table unit tests.
+                let last_value = *column.last().unwrap_or(&ZERO);
+                column.resize(trace_len, last_value);
```
Comment (grjte) [inline processor/src/decoder/trace.rs:333]: This doesn't sound like it should be too hard to fix. Is it worth making an issue after this is merged?

File: processor/src/decoder/tests.rs [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+    },
+    program::blocks::{CodeBlock, Span},
+    FieldElement, StarkField, DECODER_TRACE_RANGE,
+};
+
+// CONSTANTS
+// ================================================================================================
+
+const ONE: Felt = Felt::ONE;
+const ZERO: Felt = Felt::ZERO;
```
Comment (grjte) [inline processor/src/decoder/tests.rs:16]: I like the readability of just using ONE and ZERO, and I notice you've done this in 3 or 4 files. What would you think of putting this in core and doing it everywhere?

File: processor/src/decoder/tests.rs [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+fn build_trace(stack: &[u64], program: &CodeBlock) -> (DecoderTrace, usize) {
+    let inputs = ProgramInputs::new(stack, &[], vec![]).unwrap();
+    let mut process = Process::new(inputs);
+    process.execute_code_block(&program).unwrap();
```
Comment (grjte) [inline processor/src/decoder/tests.rs:549]: nit: I don't think we need this borrow. It could just be `program`.

**Comment by grjte** [comment]
I meant to ask about one other thing - do we need to have a boundary constraint on the starting row address of the hasher? I haven't thought about this in depth (and I think we're ok), but it crossed my mind as I was reviewing, and you may have thoughts since you've been looking at this deeply, so it seemed worth mentioning

**Comment by bobbinth** [comment]
> do we need to have a boundary constraint on the starting row address of the hasher?

I think we do because we are now assuming that the address starts at 1 (if the prover sets it to higher values, it's not a problem, but setting it to 0 may lead to some unintended consequences). I actually mention this in the 3rd point in the PR description above :)

**Comment by grjte** [comment]
> > do we need to have a boundary constraint on the starting row address of the hasher?
> 
> I think we do because we are now assuming that the address starts at 1 (if the prover sets it to higher values, it's not a problem, but setting it to 0 may lead to some unintended consequences). I actually mention this in the 3rd point in the PR description above :)

Ah, so you do! Sorry, I didn't reread the description after finishing the review, so it's been a couple days. Yep, that makes sense - we definitely do need the constraint.

File: processor/src/decoder/mod.rs [hunk 5]
```diff
@@ -9,107 +23,481 @@ impl Process {
...
+        // hashing operation batches. Thus, the result of the hash is expected to be in row
+        // addr + (num_batches * 8) - 1.
+        let op_batches = block.op_batches();
+        let (addr, _result) = self.hasher.hash_span_block(op_batches);
+
+        // make sure the result computed by the hasher is the same as the expected block hash
+        debug_assert_eq!(block.hash(), _result.into());
+
+        // start decoding the first operation batch; this also appends a row with SPAN operation
+        // to th decoder trace. we also need the total number of operation groups so that we can
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decoder/mod.rs:154]: Fixed!

File: processor/src/decoder/mod.rs [hunk 6]
```diff
@@ -9,107 +23,481 @@ impl Process {
...
+/// In the above, the meaning of the columns is as follows:
+/// * addr column contains address of the hasher for the current block (row index from the
+///   auxiliary hashing table). It also serves the role of unique block identifiers. This is
+///   convenient, because hasher addresses are guaranteed to be unique.
+/// * op_bits columns b0 through b6 are used to encode an operation to be executed by the VM.
+///   Each of these columns contains a single binary value, which together form a single opcode.
+/// * in_span column is a binary flag set to ONE when we are inside a SPAN block, and to ZERO
+///   otherwise.
+/// * Hasher state columns h0 through h7. These are multi purpose columns used as follows:
+///   - When starting decoding of a new code block (e.g., vai JOIN, SPLIT, LOOP, SPAN operations)
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decoder/mod.rs:194]: fixed!

File: processor/src/decoder/mod.rs [hunk 7]
```diff
@@ -9,107 +23,481 @@ impl Process {
...
+            Operation::Loop,
+            loop_body_hash,
+            [Felt::ZERO; 4],
+        );
+    }
 
-    pub fn start_split(&mut self, _block: &Split, _addr: Felt, _condition: Felt) {}
+    /// Starts decoding another iteration of a loop.
+    ///
+    /// This appending execution of a REPEAT operation to the trace.
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decoder/mod.rs:272]: fixed!

File: processor/src/decoder/mod.rs [hunk 8]
```diff
@@ -9,107 +23,481 @@ impl Process {
...
+    pub fn repeat(&mut self) {
+        let block_info = self.block_stack.peek();
+        debug_assert_eq!(Felt::ONE, block_info.is_loop);
+        self.trace.append_loop_repeat(block_info.addr);
+    }
 
-    pub fn end_split(&mut self, _block: &Split) {}
+    /// Ends decoding of a control block (i.e., a non-SPAN block).
+    ///
+    /// This appending execution of an END operation to the trace. The top block on the block
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decoder/mod.rs:281]: Fixed!

File: core/src/hasher/mod.rs [hunk 9]
```diff
@@ -103,3 +110,67 @@ pub fn apply_round(state: &mut [Felt; STATE_WIDTH], round: usize) {
...
+        init_values[3],
+        init_values[4],
+        init_values[5],
+        init_values[6],
+        init_values[7],
+    ]
+}
+
+/// Initializes hasher state with the elements from the provided words. The number of elements
+/// to be hashes is set to 8.
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply core/src/hasher/mod.rs:138]: fixed!

File: core/src/hasher/mod.rs [hunk 10]
```diff
@@ -103,3 +110,67 @@ pub fn apply_round(state: &mut [Felt; STATE_WIDTH], round: usize) {
...
+    state[5] += values[1];
+    state[6] += values[2];
+    state[7] += values[3];
+    state[8] += values[4];
+    state[9] += values[5];
+    state[10] += values[6];
+    state[11] += values[7];
+}
+
+/// Returns elements representing the digest portion of the provide hasher's state.
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply core/src/hasher/mod.rs:171]: Fixed!

File: processor/src/hasher/mod.rs [hunk 11]
```diff
@@ -90,16 +92,72 @@ impl Hasher {
...
+
+    /// Computes a sequential hash of all operation batches in the list and returns the result.
+    ///
+    /// The returned tuple also contains the row address of the execution trace at which hash
+    /// computation started.
+    pub fn hash_span_block(&mut self, op_batches: &[OpBatch]) -> (Felt, Word) {
+        const START: Selectors = LINEAR_HASH;
+        const RETURN: Selectors = RETURN_HASH;
+        // absorb selectors are the same as linear hash selectors, but absorb selectors are
+        // applies on the last row of a permutation cycle, while linear hash selectors are
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/hasher/mod.rs:121]: fixed!

File: processor/src/decoder/trace.rs [hunk 12]
```diff
@@ -0,0 +1,374 @@
...
+        let last_group_count = self.last_group_count();
+        debug_assert!(last_group_count == ZERO, "group count not zero");
+        self.group_count_trace.push(last_group_count);
+
+        self.op_idx_trace.push(ZERO);
+    }
+
+    /// Appends a trace row marking the beginning of a new loop iteration.
+    ///
+    /// When we starting a new loop iteration, we do the following:
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decoder/trace.rs:137]: Fixed!

File: processor/src/decoder/trace.rs [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+            self.hasher_trace[i].push(op_group);
+        }
+        self.group_count_trace.push(num_op_groups);
+        self.op_idx_trace.push(ZERO);
+    }
+
+    /// Appends a trace row marking a RESPAN operation.
+    ///
+    /// When a RESPAN operation is executed, we do the following:
+    /// - Copy over the block address from the previous row. The SPAN address will be update in
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decoder/trace.rs:190]: Fixed!

File: processor/src/decoder/trace.rs [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+    /// - Set op_bits to the opcode of the executed operation.
+    /// - Set is_span to ONE.
+    /// - Set the first hasher state register to the aggregation of remaining operations to be
+    ///   executed in the current operation group.
+    /// - Set the second hasher state register to the address of the SPAN's parent block.
+    /// - Set the remaining hasher state registers to ZEROs.
+    /// - Set the number of groups remaining to be processed. This number of groups changes if
+    ///   in the previous row an operation with an immediate value was executed or if this
+    ///   operation is a start of a new operation group.
+    /// - Set the operation's index withing the current operation group.
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decoder/trace.rs:221]: fixed!

File: processor/src/decoder/trace.rs [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+            "target trace length too small"
+        );
+
+        let mut trace = Vec::new();
+
+        // put ZEROs into the unfilled rows of block address column
+        self.addr_trace.resize(trace_len, ZERO);
+        trace.push(self.addr_trace);
+
+        // insert HALT opcode into the unfilled rows of ob_bits columns
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decoder/trace.rs:310]: Fixed!

File: processor/src/decoder/tests.rs [hunk 16]
```diff
@@ -0,0 +1,675 @@
...
+    trace: &DecoderTrace,
+    row_idx: usize,
+    addr: Felt,
+    op: Operation,
+    group_count: u64,
+    op_idx: u64,
+    in_span: u64,
+) {
+    assert_eq!(trace[ADDR_COL_IDX][row_idx], addr);
+    assert!(contains_op(&trace, row_idx, op));
```
[the remark this answers is not in the document]
Comment (bobbinth) [reply processor/src/decoder/tests.rs:576]: Fixed!

File: processor/src/decoder/tests.rs (hunk 4 above)
Comment (bobbinth) [reply processor/src/decoder/tests.rs:549]: Fixed!

File: processor/src/decoder/tests.rs (hunk 3 above)
Comment (bobbinth) [reply processor/src/decoder/tests.rs:16]: Yes - good idea! But I'll leave it for future PRs.

File: processor/src/decoder/trace.rs (hunk 2 above)
Comment (bobbinth) [reply processor/src/decoder/trace.rs:333]: Yes - but it is not trivial either. So, yes - let's create an issue for this.

File: core/src/hasher/mod.rs (hunk 1 above)
Comment (bobbinth) [reply core/src/hasher/mod.rs:52]: Agreed! I renamed it and updated relevant references.

**Pull request merged**